### PR TITLE
[WIP,Feature] Add cudagraphs and compile option

### PR DESCRIPTION
**Pull request by vmoens**

### CODE REVIEW

**Comment by vmoens** [comment]
I should revert these tabs :p

File: docker/environment.yaml
Comment (vmoens) [inline docker/environment.yaml:1]: This could be dropped, but conda was hanging forever to create the env so I ended up creating my own requirements.txt

File: tdmpc2/common/buffer.py [hunk 1]
```diff
@@ -12,7 +12,7 @@ class Buffer():
 
 	def __init__(self, cfg):
 		self.cfg = cfg
-		self._device = torch.device('cuda')
```
Comment (vmoens) [inline tdmpc2/common/buffer.py:15]: always safer to use ordinals - because `if self._device == tensor.device` could fail otherwise

File: tdmpc2/common/buffer.py [hunk 2]
```diff
@@ -41,8 +41,8 @@ def _reserve_buffer(self, storage):
 		return ReplayBuffer(
 			storage=storage,
 			sampler=self._sampler,
-			pin_memory=True,
-			prefetch=1,
```
Comment (vmoens) [inline tdmpc2/common/buffer.py:45]: disabling prefetch bc I haven't tested how multithreaded cuda ops work while executing a cudagraph.
Also pin_memory incur some overhead, best to avoid it
https://docs-preview.pytorch.org/pytorch/tutorials/2983/intermediate/pinmem_nonblock.html

File: tdmpc2/common/buffer.py [hunk 3]
```diff
@@ -58,7 +58,7 @@ def _init(self, tds):
 		total_bytes = bytes_per_step*self._capacity
 		print(f'Storage required: {total_bytes/1e9:.2f} GB')
 		# Heuristic: decide whether to use CUDA or CPU memory
-		storage_device = 'cuda' if 2.5*total_bytes < mem_free else 'cpu'
```
Comment (vmoens) [inline tdmpc2/common/buffer.py:61]: ditto

File: tdmpc2/common/buffer.py [hunk 4]
```diff
@@ -92,5 +92,5 @@ def add(self, td):
 
 	def sample(self):
 		"""Sample a batch of subsequences from the buffer."""
-		td = self._buffer.sample().view(-1, self.cfg.horizon+1).permute(1, 0)
+		td = self._buffer.sample().view(-1, self.cfg.horizon+1).permute(1, 0).clone()
```
Comment (vmoens) [inline tdmpc2/common/buffer.py:95]: because we're doing views but we'll feed this to the model, better cloning the values (to make sure we don't have a view on CUDA)
This could be handled by CudaGraphModule btw! I will patch it

File: tdmpc2/common/layers.py [hunk 5]
```diff
@@ -32,13 +32,13 @@ class ShiftAug(nn.Module):
 	def __init__(self, pad=3):
 		super().__init__()
 		self.pad = pad
+		self.padding = tuple([self.pad] * 4)
```
Comment (vmoens) [inline tdmpc2/common/layers.py:35]: quality: I don't think we should re-compute this every time. Negligeable overhead but still...

File: tdmpc2/common/layers.py [hunk 6]
```diff
@@ -32,13 +32,13 @@ class ShiftAug(nn.Module):
...
 
 	def forward(self, x):
 		x = x.float()
 		n, _, h, w = x.size()
-		assert h == w
-		padding = tuple([self.pad] * 4)
-		x = F.pad(x, padding, 'replicate')
+		# assert h == w
```
Comment (vmoens) [inline tdmpc2/common/layers.py:40]: no worry with the assert, could be set backl

File: tdmpc2/common/layers.py [hunk 7]
```diff
@@ -59,7 +59,7 @@ def __init__(self):
 		super().__init__()
 
 	def forward(self, x):
-		return x.div_(255.).sub_(0.5)
+		return x.div(255.).sub(0.5)
```
Comment (vmoens) [inline tdmpc2/common/layers.py:62]: avoid in-place ops in PyTorch unless absolutely necessary (eg, updating params of a nn.Module).
I believe compile undoes the in-place and contrary to common belief in-place is rarely faster than out-place

File: tdmpc2/common/math.py [hunk 8]
```diff
@@ -9,19 +9,19 @@ def soft_ce(pred, target, cfg):
...
+
 def _gaussian_residual(eps, log_std):
 	return -0.5 * eps.pow(2) - log_std
 
 
-@torch.jit.script
+LOG2PI = 1.8378770351409912
 def _gaussian_logprob(residual):
-	return residual - 0.5 * torch.log(2 * torch.pi)
+	return residual - 0.5 * LOG2PI
```
Comment (vmoens) [inline tdmpc2/common/math.py:24]: this was breaking in eager (when removing the jit)
Also jit + compile is some uncharted territory. I'm removing them all

File: tdmpc2/common/math.py [hunk 9]
```diff
@@ -72,9 +72,9 @@ def two_hot(x, cfg):
 	x = torch.clamp(symlog(x), cfg.vmin, cfg.vmax).squeeze(1)
 	bin_idx = torch.floor((x - cfg.vmin) / cfg.bin_size).long()
 	bin_offset = ((x - cfg.vmin) / cfg.bin_size - bin_idx.float()).unsqueeze(-1)
-	soft_two_hot = torch.zeros(x.size(0), cfg.num_bins, device=x.device)
-	soft_two_hot.scatter_(1, bin_idx.unsqueeze(1), 1 - bin_offset)
-	soft_two_hot.scatter_(1, (bin_idx.unsqueeze(1) + 1) % cfg.num_bins, bin_offset)
+	soft_two_hot = x.new_zeros(x.shape[0], cfg.num_bins)
+	soft_two_hot = soft_two_hot.scatter(1, bin_idx.unsqueeze(1), 1 - bin_offset)
+	soft_two_hot = soft_two_hot.scatter(1, (bin_idx.unsqueeze(1) + 1) % cfg.num_bins, bin_offset)
```
Comment (vmoens) [inline tdmpc2/common/math.py:77]: same: avoid in-place ops. `new_zeros` isn't very different than the `zeros` before, just avoid reading the device of x directly.

File: tdmpc2/common/scale.py [hunk 10]
```diff
@@ -18,7 +18,7 @@ def load_state_dict(self, state_dict):
 
 	@property
 	def value(self):
-		return self._value.cpu().item()
+		return self._value
```
Comment (vmoens) [inline tdmpc2/common/scale.py:21]: avoid `cpu()` and `item()` which will not work with cudagraphs
`item()` causes a cuda synchronization, which is bad, and I think it also causes a graph break.

File: tdmpc2/common/world_model.py [hunk 11]
```diff
@@ -160,13 +160,14 @@ def Q(self, z, a, task, return_type='min', target=False):
...
+
 		z = torch.cat([z, a], dim=-1)
 		out = (self._target_Qs if target else self._Qs)(z)
 
 		if return_type == 'all':
 			return out
 
-		Q1, Q2 = out[np.random.choice(self.cfg.num_q, 2, replace=False)]
+		rand_idx = torch.randperm(self.cfg.num_q)[:2].unbind(0)
+		Q1, Q2 = out[rand_idx[0]], out[rand_idx[1]]
```
Comment (vmoens) [inline tdmpc2/common/world_model.py:171]: just a PT version of the above. Cudagraphs didn't like the auto-unbind like `Q1, Q2 = out[idx]` so I'm doing that manually

File: tdmpc2/tdmpc2.py [hunk 12]
```diff
@@ -23,16 +23,32 @@ def __init__(self, cfg):
 			{'params': self.model._dynamics.parameters()},
 			{'params': self.model._reward.parameters()},
 			{'params': self.model._Qs.parameters()},
-			{'params': self.model._task_emb.parameters() if self.cfg.multitask else []}
-		], lr=self.cfg.lr)
-		self.pi_optim = torch.optim.Adam(self.model._pi.parameters(), lr=self.cfg.lr, eps=1e-5)
+			{'params': self.model._task_emb.parameters() if self.cfg.multitask else []
+			 }
+		], lr=self.cfg.lr, capturable=True)
```
Comment (vmoens) [inline tdmpc2/tdmpc2.py:28]: capturable makes it possible to use cudagraphs without compile. With compile it's not necessary

File: tdmpc2/tdmpc2.py [hunk 13]
```diff
@@ -23,16 +23,32 @@ def __init__(self, cfg):
...
 		self.discount = torch.tensor(
 			[self._get_discount(ep_len) for ep_len in cfg.episode_lengths], device='cuda'
 		) if self.cfg.multitask else self._get_discount(cfg.episode_length)
 
+		self.plan_dict = {
+				(True, True): functools.partial(self.plan, t0=True, eval_mode=True),
+				(False, True): functools.partial(self.plan, t0=False, eval_mode=True),
+				(True, False): functools.partial(self.plan, t0=True, eval_mode=False),
+				(False, False): functools.partial(self.plan, t0=False, eval_mode=False),
+				}
```
Comment (vmoens) [inline tdmpc2/tdmpc2.py:42]: this isn't very elegant but we need to have a separate cudagraph for each `plan` config (unless a tensor is passed).
We could automate that in CudaGraphModule but that would induce some overhead (check the non-tensor inputs). So I'm doing that manually here 
Not terribly sexy, happy to hear thoughts about this!

File: tdmpc2/tdmpc2.py [hunk 14]
```diff
@@ -86,7 +102,10 @@ def act(self, obs, t0=False, eval_mode=False, task=None):
 			task = torch.tensor([task], device=self.device)
 		z = self.model.encode(obs, task)
 		if self.cfg.mpc:
-			a = self.plan(z, t0=t0, eval_mode=eval_mode, task=task)
+			kwargs = {}
+			if task is not None:
+				kwargs["task"] = task
```
Comment (vmoens) [inline tdmpc2/tdmpc2.py:107]: here too, plan won't like non-tensor args so if task is None I just don't pass it

File: tdmpc2/tdmpc2.py [hunk 15]
```diff
@@ -162,18 +181,19 @@ def plan(self, z, t0=False, eval_mode=False, task=None):
 				std = std * self.model._action_masks[task]
 
 		# Select action
-		score = score.squeeze(1).cpu().numpy()
-		actions = elite_actions[:, np.random.choice(np.arange(score.shape[0]), p=score)]
+		score = score.squeeze(1)
```
Comment (vmoens) [inline tdmpc2/tdmpc2.py:184]: calling cpu() induces another sync, everything here can be done on cuda (and fused using compile / recorded using cudagraphs)

File: tdmpc2/tdmpc2.py [hunk 16]
```diff
@@ -162,18 +181,19 @@ def plan(self, z, t0=False, eval_mode=False, task=None):
...
+		rand_idx = torch.multinomial(score, 1)
+		actions = elite_actions[:, rand_idx].squeeze(1)
 		self._prev_mean = mean
 		a, std = actions[0], std[0]
 		if not eval_mode:
-			a += std * torch.randn(self.cfg.action_dim, device=std.device)
-		return a.clamp_(-1, 1)
-		
+			a = a + std * torch.randn(self.cfg.action_dim, device=std.device)
```
Comment (vmoens) [inline tdmpc2/tdmpc2.py:190]: no inplace

File: tdmpc2/tdmpc2.py [hunk 17]
```diff
@@ -162,18 +181,19 @@ def plan(self, z, t0=False, eval_mode=False, task=None):
...
+		return a.clamp(-1, 1)
```
Comment (vmoens) [inline tdmpc2/tdmpc2.py:191]: ditto

File: tdmpc2/tdmpc2.py [hunk 18]
```diff
@@ -218,48 +238,53 @@ def _td_target(self, next_z, reward, task):
...
+
 		Returns:
 			dict: Dictionary of training statistics.
 		"""
 		obs, action, reward, task = buffer.sample()
-	
+		kwargs = {}
+		if task is not None:
+			kwargs["task"] = task
+		return self._update(obs, action, reward, **kwargs)
```
Comment (vmoens) [inline tdmpc2/tdmpc2.py:252]: we can't compile / record buffers yet :/
I'm working on this but there are some loose ends that I'm struggling to solve (recompile at each extension for instance), so I'm taking it away from the compiled code

File: tdmpc2/tdmpc2.py [hunk 19]
```diff
@@ -218,48 +238,53 @@ def _td_target(self, next_z, reward, task):
...
+
+	def _update(self, obs, action, reward, task=None):
 		# Compute targets
 		with torch.no_grad():
 			next_z = self.model.encode(obs[1:], task)
 			td_targets = self._td_target(next_z, reward, task)
 
 		# Prepare for update
-		self.optim.zero_grad(set_to_none=True)
+		self.optim.zero_grad()
```
Comment (vmoens) [inline tdmpc2/tdmpc2.py:261]: just not sure how that plays with compile - should check! better to be safe (?)

File: tdmpc2/tdmpc2.py [hunk 20]
```diff
@@ -218,48 +238,53 @@ def _td_target(self, next_z, reward, task):
...
 		zs = torch.empty(self.cfg.horizon+1, self.cfg.batch_size, self.cfg.latent_dim, device=self.device)
 		z = self.model.encode(obs[0], task)
 		zs[0] = z
 		consistency_loss = 0
-		for t in range(self.cfg.horizon):
-			z = self.model.next(z, action[t], task)
-			consistency_loss += F.mse_loss(z, next_z[t]) * self.cfg.rho**t
+		for t, (_action, _next_z) in enumerate(zip(action.unbind(0), next_z.unbind(0))):
+			z = self.model.next(z, _action, task)
+			consistency_loss += F.mse_loss(z, _next_z) * self.cfg.rho**t
```
Comment (vmoens) [inline tdmpc2/tdmpc2.py:271]: here and below: I prefer unbind rather than indexing, it's usually faster (bc you access the original tensor just once to slice it)
but it should work with compile and cudagraphs without problems I think

File: tdmpc2/tdmpc2.py [hunk 21]
```diff
@@ -218,48 +238,53 @@ def _td_target(self, next_z, reward, task):
...
-			reward_loss += math.soft_ce(reward_preds[t], reward[t], self.cfg).mean() * self.cfg.rho**t
-			for q in range(self.cfg.num_q):
-				value_loss += math.soft_ce(qs[q][t], td_targets[t], self.cfg).mean() * self.cfg.rho**t
-		consistency_loss *= (1/self.cfg.horizon)
-		reward_loss *= (1/self.cfg.horizon)
-		value_loss *= (1/(self.cfg.horizon * self.cfg.num_q))
+		for t, (_rew_pred, _rew, _td_targets) in enumerate(zip(reward_preds.unbind(0), reward.unbind(0), td_targets.unbind(0))):
+			reward_loss = reward_loss + math.soft_ce(_rew_pred, _rew, self.cfg).mean() * self.cfg.rho**t
+			for q, _qs in enumerate(qs.unbind(0)):
+				value_loss = value_loss + math.soft_ce(_qs[t], _td_targets, self.cfg).mean() * self.cfg.rho**t
```
Comment (vmoens) [inline tdmpc2/tdmpc2.py:284]: ditto

File: tdmpc2/tdmpc2.py [hunk 22]
```diff
@@ -218,48 +238,53 @@ def _td_target(self, next_z, reward, task):
...
+		consistency_loss /= self.cfg.horizon
+		reward_loss /= self.cfg.horizon
+		value_loss /= self.cfg.horizon * self.cfg.num_q
```
Comment (vmoens) [inline tdmpc2/tdmpc2.py:287]: just reducing the number of ops and making things a bit more elegant

File: tdmpc2/tdmpc2.py [hunk 23]
```diff
@@ -279,12 +304,12 @@ def update(self, buffer):
...
+			"value_loss": value_loss,
 			"pi_loss": pi_loss,
-			"total_loss": float(total_loss.mean().item()),
-			"grad_norm": float(grad_norm),
-			"pi_scale": float(self.scale.value),
-		}
+			"total_loss": total_loss,
+			"grad_norm": grad_norm,
+			"pi_scale": self.scale.value,
+		}).detach().mean()
```
Comment (vmoens) [inline tdmpc2/tdmpc2.py:315]: we need to return only detached tensors
I rely on tensordict's mean and detach for this

File: tdmpc2/trainer/online_trainer.py [hunk 24]
```diff
@@ -3,7 +3,7 @@
 import numpy as np
 import torch
 from tensordict.tensordict import TensorDict
-
+from torchrl._utils import timeit
```
Comment (vmoens) [inline tdmpc2/trainer/online_trainer.py:6]: just to quickly benchmark the time it takes to execute update and plan

**Comment by nicklashansen** [comment]
@vmoens I updated `docker/environment.yaml` to use nightly versions. Worked for me!
